arangod/RocksDBEngine/RocksDBMethods.h

```diff
@@ -77,67 +78,98 @@ class RocksDBMethods {
 // THIS IS BECAUSE THESE TRANSACTIONS WILL EITHER READ FROM
 // OR (XOR) WRITE TO A COLLECTION. IF THIS PRECONDITION IS
 // VIOLATED THE DISABLED INDEXING WILL BREAK GET OPERATIONS.
 struct IndexingDisabler {
-  // will only be active if condition is true
-
-  IndexingDisabler() = delete;
-  IndexingDisabler(IndexingDisabler&&) = delete;
   IndexingDisabler(IndexingDisabler const&) = delete;
   IndexingDisabler& operator=(IndexingDisabler const&) = delete;
-  IndexingDisabler& operator=(IndexingDisabler&&) = delete;
 
-  IndexingDisabler(RocksDBMethods* meth, bool condition) : _meth(nullptr) {
+  // will only be active if condition is true
+  IndexingDisabler(RocksDBMethods* methods, bool condition)
+      : _methods(nullptr) {
+    TRI_ASSERT(methods != nullptr);
+
     if (condition) {
-      bool disabledHere = meth->DisableIndexing();
+      bool disabledHere = methods->DisableIndexing();
       if (disabledHere) {
-        _meth = meth;
+        _methods = methods;
       }
     }
   }
 
   ~IndexingDisabler() {
-    if (_meth) {
-      _meth->EnableIndexing();
+    if (_methods != nullptr) {
+      _methods->EnableIndexing();
     }
   }
 
  private:
-  RocksDBMethods* _meth;
+  RocksDBMethods* _methods;
 };
 
 // if only single indices should be enabled during operations
 struct IndexingEnabler {
-  // will only be active if condition is true
-
-  IndexingEnabler() = delete;
-  IndexingEnabler(IndexingEnabler&&) = delete;
   IndexingEnabler(IndexingEnabler const&) = delete;
   IndexingEnabler& operator=(IndexingEnabler const&) = delete;
-  IndexingEnabler& operator=(IndexingEnabler&&) = delete;
 
-  IndexingEnabler(RocksDBMethods* meth, bool condition) : _meth(nullptr) {
+  // will only be active if condition is true
+  IndexingEnabler(RocksDBMethods* methods, bool condition) : _methods(nullptr) {
+    TRI_ASSERT(methods != nullptr);
+
     if (condition) {
-      bool enableHere = meth->EnableIndexing();
+      bool enableHere = methods->EnableIndexing();
       if (enableHere) {
-        _meth = meth;
+        _methods = methods;
       }
     }
   }
 
   ~IndexingEnabler() {
-    if (_meth) {
-      _meth->DisableIndexing();
+    if (_methods != nullptr) {
+      _methods->DisableIndexing();
+    }
+  }
+
+ private:
+  RocksDBMethods* _methods;
+};
+
+class ConcurrencyControlToggler {
+ public:
+  ConcurrencyControlToggler(ConcurrencyControlToggler const&) = delete;
+  ConcurrencyControlToggler& operator=(ConcurrencyControlToggler const&) =
+      delete;
+
+  explicit ConcurrencyControlToggler(RocksDBMethods* methods, bool isExclusive)
+      : _methods(methods), _exclusive(isExclusive) {
+    TRI_ASSERT(methods != nullptr);
+  }
+
+  ~ConcurrencyControlToggler() {
+    // nothing to do for exclusive transactions, as they already have the
+    // concurrency control disabled for the entire transaction
+    if (!_exclusive) {
+      _methods->SetSkipConcurrencyControl(false);
+    }
+  }
+
+  void setConcurrencyControl(bool value) {
```

Comment:
**NIT**
Why do you decide to change `SetSkipConcurrencyControl` to opposite `setConcurrencyControl`?
Anyway I think `skipConcurrencyControl` is better option.

Reply:
The name of the member inside RocksDB's PessimisticTransaction is also `skip_concurrency_control_`.
And its value defaults to `false`, so it is only called in exceptional situations. So the thing here is not to enable the concurrency control, but rather to disable it (temporarily).

Reply:
Anyway, I made changes to the PR as a whole now.